### ISSUE TITLE
Route reuse strategy throwing RangeError: Maximum call stack size exceeded in Angular 13

### ISSUE THREAD

**Issue by ChabuRah** [angular/angular#44383]
### Which @angular/* package(s) are the source of the bug?

router

### Is this a regression?

No

### Description

We have a considerable large angular app, where we use custom route reuse strategy for performance gains.
The route reuse strategy we use was working perfectly fine in angular 12. However, when we upgraded to Angular 13 we have started seeing the following error:

`Error: Uncaught (in promise): RangeError: Maximum call stack size exceeded`

The error happens in this scenario:
- Go to Route A
- Go to Route B
- Go to Route A again(either by hitting browser back, or by going to the route A through href on <a> tag on page B), crash and navigation doesn't happen.

The inifinite loop is happening inside Angular router's code (router.mjs) in RouterStateSnapshot's constructor, that calls into following function:

`function setRouterState(state, node) {
    node.value._routerState = state;
    node.children.forEach(c => setRouterState(state, c));
}`

Here is how we have our custom route reuse strategy:

```
import {
	RouteReuseStrategy,
	DetachedRouteHandle,
	ActivatedRouteSnapshot,
} from "@angular/router";

export class CustomRouteReuseStrategy implements RouteReuseStrategy {
	private handlers: { [key: string]: DetachedRouteHandle } = {};

	constructor() {}

	public shouldDetach(_route: ActivatedRouteSnapshot): boolean {
		return true;
	}

	public store(
		route: ActivatedRouteSnapshot,
		handle: DetachedRouteHandle
	): void {
		this.handlers[route.data.title] = handle;
	}

	public shouldAttach(route: ActivatedRouteSnapshot): boolean {
		return !!this.handlers[route.data.title];
	}

	public retrieve(route: ActivatedRouteSnapshot): DetachedRouteHandle {
		return this.handlers[route.data.title];
	}

	public shouldReuseRoute(
		future: ActivatedRouteSnapshot,
		curr: ActivatedRouteSnapshot
	): boolean {
		const reuse = future.routeConfig === curr.routeConfig;
		return reuse;
	}
}
```

Here is how we add route to Routing Module:

`
	{
		path: "post/:id",
		loadChildren: () =>
			import("path-to-post-module").then(
				(module) => module.PostModule
			),
		data: { title: "post" },
	},`

This may sound like a stack overflow question, but this setup used to work in Angular 12 without any issues. So, I'm not sure what changed in version 13, that's causing this crash. If I remove the route reuse strategy, this error doesn't happen, but obviously we want the route reuse in our app.

### Please provide a link to a minimal reproduction of the bug
https://github.com/ChabuRah/angular-router-error

### Please provide the exception or error you saw

```true
Error: Uncaught (in promise): RangeError: Maximum call stack size exceeded
RangeError: Maximum call stack size exceeded
    at setRouterState (router.mjs:1610)
    at router.mjs:1612
    at Array.forEach (<anonymous>)
    at setRouterState (router.mjs:1612)
    at router.mjs:1612
    at Array.forEach (<anonymous>)
    at setRouterState (router.mjs:1612)
    at router.mjs:1612
    at Array.forEach (<anonymous>)
    at setRouterState (router.mjs:1612)
    at resolvePromise (zone.js:1213)
    at resolvePromise (zone.js:1167)
    at zone.js:1279
    at ZoneDelegate.invokeTask (zone.js:406)
    at Object.onInvokeTask (core.mjs:25463)
    at ZoneDelegate.invokeTask (zone.js:405)
    at Zone.runTask (zone.js:178)
    at drainMicroTaskQueue (zone.js:582)
    at invokeTask (zone.js:491)
    at ZoneTask.invoke (zone.js:476)
```

### Please provide the environment you discovered this bug in (run `ng version`)

```true
Angular CLI: 13.0.3
Node: 14.15.5
Package Manager: npm 6.14.11
OS: linux x64

Package                      Version
------------------------------------------------------
@angular-devkit/architect    0.1300.3 (cli-only)
@angular-devkit/core         13.0.3 (cli-only)
@angular-devkit/schematics   13.0.3 (cli-only)
@schematics/angular          13.0.3 (cli-only)
```

**Comment by JoostK** [comment]
Could you please create a minimum reproduction of the issue you're seeing so we can investigate.

**Comment by Manndung75** [comment]
A

**Comment by ChabuRah** [comment]
Here is the repro: https://github.com/ChabuRah/angular-router-error.
Readme has instructions on how to get to the crash, Appreciate your response!

**Comment by blasco** [comment]
I´m experiencing the exact same behavior

**Comment by JoostK** [comment]
I suspect this broke in #43791.

The reuse strategy uses a route's `data` property to store/retrieve an existing route handler. However, route data is inherited from parent routes, such that both the lazy route configuration with `data: { title: '...' }` and its component-routes within the lazy-loaded module share the same title. This poses a conflict for the reuse strategy, which now finds the parent route handle when reloading the child, resulting in an infinite loop.

I'll defer to @atscott as to what the solution could be here.

**Comment by ChabuRah** [comment]
Thanks @JoostK for looking into it.

@atscott We have put our update on hold due to this issue, so I'll be more than happy to know if there is any workaround until the fix is out.

**Comment by atscott** [comment]
@JoostK is correct that #43791 would be the cause for this failure. We should add an item to the breaking changes notes. Joost mentions that the child retrieves the parent route handle, which results in the infinite loop. This is a symptom of the implementation of this particular reuse strategy and not necessarily a bug in the Router implementation. 

#43791 addressed an issue where a parent with more than one child would not be able to correctly navigate between child routes when the parent route was stored. To fix this, we need to not only store/attach a parent, but also store/attach its children. If the implementation of the route reuse strategy returns the same value for both a parent and a child then the state becomes an infinite loop.

In the provided reproduction, the route reuse strategy _could_ be updated to 

```
public shouldAttach(route: ActivatedRouteSnapshot): boolean {
		return !!this.handlers[route.data.title] && !!route.component;
	}

	public retrieve(route: ActivatedRouteSnapshot): DetachedRouteHandle|null {
		if (!route.component) return null;
		return this.handlers[route.data.title];
	}
```

In short, the `RouteReuseStrategy` needs to be updated to not retrieve the same handle for both a parent and a child. The update above prevents the parent `loadChildren` config from being stored/retrieved at all.

So the action items here are:

* Add a breaking changes note to the changelog
* Potentially attempt to detect and prevent these infinite loops and add a debug only warning.

**Comment by ChabuRah** [comment]
@atscott the workaround does't seem to work.
I'm running into another crash after applying your workaround:

```
Error: Uncaught (in promise): TypeError: Cannot read properties of null (reading 'contexts')
TypeError: Cannot read properties of null (reading 'contexts')
    at ActivateRoutes.activateRoutes (router.mjs:2140)
    at router.mjs:2110
    at Array.forEach (<anonymous>)
    at ActivateRoutes.activateChildRoutes (router.mjs:2109)
    at ActivateRoutes.activateRoutes (router.mjs:2167)
    at router.mjs:2110
    at Array.forEach (<anonymous>)
    at ActivateRoutes.activateChildRoutes (router.mjs:2109)
    at ActivateRoutes.activate (router.mjs:2023)
    at router.mjs:2008
    at resolvePromise (zone.js:1213)
    at resolvePromise (zone.js:1167)
    at zone.js:1279
    at ZoneDelegate.invokeTask (zone.js:406)
    at Object.onInvokeTask (core.mjs:25463)
    at ZoneDelegate.invokeTask (zone.js:405)
    at Zone.runTask (zone.js:178)
    at drainMicroTaskQueue (zone.js:582)
```
When I put the breakpoints it's crashing at this line:
```
else {
            if (future.component) {
                // if we have a normal route, we need to place the component into the outlet and recurse.
                const context = parentContexts.getOrCreateContext(future.outlet);
                if (this.routeReuseStrategy.shouldAttach(future.snapshot)) {
                    const stored = this.routeReuseStrategy.retrieve(future.snapshot);
                    this.routeReuseStrategy.store(future.snapshot, null);
                    context.children.onOutletReAttached(stored.contexts); // crashes here
                    context.attachRef = stored.componentRef;
                    context.route = stored.route.value;
```

Also, looks like the update to docs merge request will close this issue. Does it mean that there won't be a fix for this in upcoming releases?

**Comment by atscott** [comment]
@ChabuRah Given the information provided and the reproduction, this does not appear to be a bug in the Router implementation so there's no fix to implement in an upcoming release.

The change I suggested may need to be modified to suit your application's specific use-case. It was only meant to work for the reproduction you provided.

**Comment by ChabuRah** [comment]
@atscott The workaround isn't working in the reproduction app as well. And if there won't be fix for this use case, then I doubt what options I've to support RouteReuseStrategy, as my use case looks like very common one.
I really appreciate your time for responding to this issue.

**Comment by atscott** [comment]
@ChabuRah The above solution worked for the reproduction I had downloaded. I'm not quite sure what's different. Anyhow, you'd likely be better off having a `Map` of `Route` to `DetachedRouteHandle` to avoid conflicts where `title` is the same. Try this:

```
export class CustomRouteReuseStrategy implements RouteReuseStrategy { 
	private handlers: Map<Route, DetachedRouteHandle> = new Map();

	constructor() {}

	public shouldDetach(_route: ActivatedRouteSnapshot): boolean {
		return true;
	}

	public store(
		route: ActivatedRouteSnapshot,
		handle: DetachedRouteHandle
	): void {
		if (!route.routeConfig) return;
		this.handlers.set(route.routeConfig, handle);
	}

	public shouldAttach(route: ActivatedRouteSnapshot): boolean {
		return !!route.routeConfig && !!this.handlers.get(route.routeConfig);
	}

	public retrieve(route: ActivatedRouteSnapshot): DetachedRouteHandle|null {
		if (!route.routeConfig || !this.handlers.has(route.routeConfig)) return null;
		return this.handlers.get(route.routeConfig)!;
	}

	public shouldReuseRoute(
		future: ActivatedRouteSnapshot,
		curr: ActivatedRouteSnapshot
	): boolean {
		return future.routeConfig === curr.routeConfig;
	}
}
```

Let me know if that doesn't work for you either.

**Comment by ChabuRah** [comment]
@atscott using routeConfig as a key seems to have worked. I think we may not need `data.title` in the route reuse strategy if the routeConfig is working as expected.
Thanks a lot for quick turnaround.
